lldb-eval/parser.cc

```diff
@@ -2709,25 +2709,69 @@ ExprResult Parser::BuildTernaryOp(ExprResult cond, ExprResult lhs,
         location, result_type, std::move(cond), std::move(lhs), std::move(rhs));
   }
 
-  // If one operand is a pointer and the other one is arithmeric, convert
-  // arithmetic operand to a pointer.
-  if ((lhs_type.IsPointerType() || lhs_type.IsNullPtrType()) &&
-      (rhs_type.IsScalarOrUnscopedEnum() || rhs_type.IsNullPtrType())) {
+  // If one operand is nullptr and the other one is literal zero, convert
+  // the literal zero to a nullptr type.
+  if (lhs_type.IsNullPtrType() && rhs->is_literal_zero()) {
     rhs = std::make_unique<CStyleCastNode>(
-        rhs->location(), lhs_type, std::move(rhs), CStyleCastKind::kPointer);
+        rhs->location(), lhs_type, std::move(rhs), CStyleCastKind::kNullptr);
 
     return std::make_unique<TernaryOpNode>(location, lhs_type, std::move(cond),
                                            std::move(lhs), std::move(rhs));
   }
-  if ((rhs_type.IsPointerType() || rhs_type.IsNullPtrType()) &&
-      (lhs_type.IsScalarOrUnscopedEnum() || lhs_type.IsNullPtrType())) {
+  if (rhs_type.IsNullPtrType() && lhs->is_literal_zero()) {
     lhs = std::make_unique<CStyleCastNode>(
-        lhs->location(), rhs_type, std::move(lhs), CStyleCastKind::kPointer);
+        lhs->location(), rhs_type, std::move(lhs), CStyleCastKind::kNullptr);
 
     return std::make_unique<TernaryOpNode>(location, rhs_type, std::move(cond),
                                            std::move(lhs), std::move(rhs));
   }
 
+  // At this point the resulting type should be a pointer, and at least one
+  // operand should either be an array or a pointer.
+
+  // Array operand types should be converted to pointer types.
+  auto lhs_type_orig = lhs_type;
+  auto rhs_type_orig = rhs_type;
+  if (lhs_type.IsArrayType()) {
+    lhs_type = lhs_type.GetArrayElementType().GetPointerType();
+  }
```

Comment:
I think we can do `array->pointer` conversion right here, i.e.

```
if (lhs_type.IsArrayType()) {
  lhs  = std::make_unique<CStyleCastNode>(...);
}
```

After this we just need to check if `lhs` and `rhs` have the same pointer type (or apply nullptr/zero -> pointer conversion).

Reply:
The reasons why I didn't convert it here are because I'd still need to do conversion for nullptr/zero later (so we have this only in one place) and to avoid allocating new memory in the case we don't have to (in the "incompatible operand types" case).

Anyway, I guess that extra allocation is negligible for performance and if you think it is more readable that way, I don't mind :)